deprecate AsdfInFits

braingram opened the pull request:
Support to reading and writing ASDF in fits is moving to [stdatamodels](https://github.com/spacetelescope/stdatamodels).

https://github.com/spacetelescope/stdatamodels/pull/110

fixes #1332

braingram commented:
@WilliamJamieson could I get your advice on handling including this deprecation?

specutils is failing because it turns warnings into errors:
https://github.com/astropy/specutils/blob/34b5e84fd107f31a3c2fe2d1aaf95517d68d95f1/setup.cfg#L61

How would you recommend we handling the error issued by this new deprecation warning?

I could open a PR with specutils ignoring the new warning (adding the following to setup.cfg):
```
ignore:ASDFInFits has been deprecated .*:asdf.exceptions.AsdfDeprecationWarning:asdf.fits_embed
```
as long as we're happy with the proposed changes in this PR. We could then wait until that PR is merged to merge this PR.

Is there an easier/better way to handle this? Thanks!

WilliamJamieson commented:
> @WilliamJamieson could I get your advice on handling including this deprecation?
> 

As far as depreciation warnings are concerned, it is not up to us to decide how `specutils` should handle them. If we want to be nice (and we do), we can open an issue there warning them that we are planning on deprecating something they are using with a timeline for when the deprecation will appear (and when the feature will be removed) alongside with recommendations on how to remediate the issue.

As far as far as actually merging this PR we should wait until `stdatamodels` has a release which contains the changes in spacetelescope/stdatamodels#110. This is because the remediation recommendation is use ASDF-in-fits from `stdatamodels`.

> specutils is failing because it turns warnings into errors: [astropy/specutils@`34b5e84`/setup.cfg#L61](https://github.com/astropy/specutils/blob/34b5e84fd107f31a3c2fe2d1aaf95517d68d95f1/setup.cfg#L61)
> 
> How would you recommend we handling the error issued by this new deprecation warning?
> 
> I could open a PR with specutils ignoring the new warning (adding the following to setup.cfg):
> 
> ```
> ignore:ASDFInFits has been deprecated .*:asdf.exceptions.AsdfDeprecationWarning:asdf.fits_embed
> ```

This is not a good idea. I would rather see a warning capture/ignore for this on a per-test basis as they will likely need to stride testing multiple version requirements.

WilliamJamieson commented on asdf/fits_embed.py:
```diff
@@ -145,6 +147,13 @@ class AsdfInFits(asdf.AsdfFile):
     """
 
     def __init__(self, hdulist=None, tree=None, **kwargs):
+        warnings.warn(
+            "AsdfInFits has been deprecated and will be removed in "
+            "asdf-3.0. Similar functionality has been added to stdatamodels "
+            "https://github.com/spacetelescope/stdatamodels to load and "
+            "save ASDF in fits files.",
+            AsdfDeprecationWarning,
+        )
```
This whole module is being deprecated, not just his class. As such the warning should be at the module level.

braingram replied:
Thanks. I moved the warning and added notes and deprecations for asdftool extract and remove_hdu

WilliamJamieson commented:
@MeeseeksDev Hello

braingram commented:
There are a few things that I'm not pleased with but require updating other packages to fix.

1) The asdf_in_fits document in stdatamodels does not appear to have a correctly formatted sphinx link (it is missing the trailing colon so it does not appear as a std:label in the object inventory):
https://raw.githubusercontent.com/spacetelescope/stdatamodels/master/docs/source/asdf_in_fits.rst
This made linking to the docs difficult. I opened a PR to fix this (https://github.com/spacetelescope/stdatamodels/pull/122) but don't think it warrants a new release.
2) The custom.css in sphinx_asdf resizes warnings when they become active:
https://github.com/asdf-format/sphinx-asdf/blob/master/sphinx_asdf/static/custom.css#L7
This makes any contained linked unclickable. So for now there is a `note` instead of a `warning` about the asdf in fits deprecation

braingram commented:
I opened a separate issue to track the deprecation of asdftool commands that require asdf.fits_embed. https://github.com/asdf-format/asdf/issues/1382

WilliamJamieson commented:
> There are a few things that I'm not pleased with but require updating other packages to fix.
> 
> 1. The asdf_in_fits document in stdatamodels does not appear to have a correctly formatted sphinx link (it is missing the trailing colon so it does not appear as a std:label in the object inventory):
>    [spacetelescope/stdatamodels@`master`/docs/source/asdf_in_fits.rst (raw)](https://raw.githubusercontent.com/spacetelescope/stdatamodels/master/docs/source/asdf_in_fits.rst)
>    This made linking to the docs difficult. I opened a PR to fix this ([fix sphinx link for asdf in fits documentation spacetelescope/stdatamodels#122](https://github.com/spacetelescope/stdatamodels/pull/122)) but don't think it warrants a new release.
> 2. The custom.css in sphinx_asdf resizes warnings when they become active:
>    [asdf-format/sphinx-asdf@`master`/sphinx_asdf/static/custom.css#L7](https://github.com/asdf-format/sphinx-asdf/blob/master/sphinx_asdf/static/custom.css?rgh-link-date=2023-02-16T21%3A23%3A11Z#L7)
>    This makes any contained linked unclickable. So for now there is a `note` instead of a `warning` about the asdf in fits deprecation

In light of these issues, I think we should wait until `sphinx_asdf` is released with fixes for these issues.

braingram commented:
> > 2. The custom.css in sphinx_asdf resizes warnings when they become active:
> >    [asdf-format/sphinx-asdf@`master`/sphinx_asdf/static/custom.css#L7](https://github.com/asdf-format/sphinx-asdf/blob/master/sphinx_asdf/static/custom.css?rgh-link-date=2023-02-16T21%3A23%3A11Z#L7)
> >    This makes any contained linked unclickable. So for now there is a `note` instead of a `warning` about the asdf in fits deprecation
> 
> In light of this issue, I think we should wait until `sphinx_asdf` is released with fixes for these issues.

I am ok with this as a note and would prefer to have it reviewed.

WilliamJamieson commented on docs/asdf/features.rst:
```diff
@@ -413,6 +413,15 @@ details).
+    will be removed from an upcoming version of ASDF. Support for reading
+    and writing ASDF data within a FITS file has moved to the `stdatamodels`
+    package. Please see the the `stdatamodels`
+    :external+stdatamodels:doc:`asdf_in_fits` documentation for information
```
This should be an intersphinx link. Is this the issue the intersphinx link that will eventually be resolved by spacetelescope/stdatamodels#122? If so we should create an issue to track updating this to a proper link when `stdatamodels` gets fixed.

braingram replied:
Thanks. I opened an issue here:
https://github.com/asdf-format/asdf/issues/1385

WilliamJamieson commented on docs/asdf/features.rst:
```diff
@@ -413,6 +413,15 @@ details).
 Saving ASDF in FITS
 ===================
 
+.. note::
+
+    `~asdf.fits_embed.AsdfInFits` and `~asdf.fits_embed` are deprecated and
+    will be removed from an upcoming version of ASDF. Support for reading
+    and writing ASDF data within a FITS file has moved to the `stdatamodels`
+    package. Please see the the `stdatamodels`
+    :external+stdatamodels:doc:`asdf_in_fits` documentation for information
+    about migrating code from ASDF to `stdatamodels`.
```
```suggestion
.. warning::

    `~asdf.fits_embed.AsdfInFits` and `~asdf.fits_embed` are deprecated and
    will be removed from an upcoming version of ASDF. Support for reading
    and writing ASDF data within a FITS file has moved to the `stdatamodels`
    package. Please see the the `stdatamodels`
    :external+stdatamodels:doc:`asdf_in_fits` documentation for information
    about migrating code from ASDF to `stdatamodels`.
```

Per https://github.com/asdf-format/asdf/pull/1337#issuecomment-1433731516 this should be a warning not a note.

braingram replied:
> I am ok with this as a note and would prefer to have it reviewed.

Changing it to a warning makes the links unclickable. I would prefer to leave this as a note.

WilliamJamieson replied:
We have gone to all the trouble to update `sphinx_asdf` to fix this issue. If you are fine with it being a note then why did we bother with all that? Especially since `sphinx_asdf` is something which is relatively easy to release a patch for.

braingram replied:
I will change it to a warning.

WilliamJamieson reviewed (changes requested):
I called out some specific points for #1382, and made some suggestions and other comments.

braingram commented on asdf/commands/extract.py:
```diff
@@ -45,6 +44,8 @@ def run(cls, args):
 
 def extract_file(input_file, output_file):
     """Function for performing extraction from ASDF-in-FITS to pure ASDF."""
+    # local import to avoid issuing Deprecation warning on every use of asdftool
+    from asdf.fits_embed import AsdfInFits
```
Yes, I figured a separate PR would be called for. If you'd rather see those changes included here let me know.

braingram commented on asdf/commands/remove_hdu.py:
```diff
@@ -39,6 +39,9 @@ def run(cls, args):
 
 def remove_hdu(input_file, output_file):
     """Function for removing ASDF HDU from ASDF-in-FITS files"""
+    # local import to trigger the deprecation warning since this command
+    # relates to and mentions AsdfInFits
+    from asdf.fits_embed import AsdfInFits  # noqa: F401
```
Yup, same as the other comment.

braingram commented on asdf/commands/tests/test_extract.py:
```diff
@@ -1,11 +1,21 @@
+    # asdf.fits_embed is imported here
+    if "asdf.fits_embed" in sys.modules:
+        del sys.modules["asdf.fits_embed"]
+    import asdf.fits_embed
+    from asdf.fits_embed import AsdfInFits
```
I added a comment to the code describing the need for this import.

In short, calling `del sys.modules['asdf.fits_embed']` isn't sufficient to allow for re-triggering the deprecation warning if an import like `from asdf.fits_embed import AsdfInFits` is used).

I chose adding the `import asdf.fits_embed` option to avoid changing the usage of AsdfInFits in test_extract.py but that should work as well (switching it to `asdf.fits_embed.AsdfInFits`). If you're prefer that alternative let me know.

WilliamJamieson replied:
See https://github.com/asdf-format/asdf/pull/1337#discussion_r1110042655.

WilliamJamieson commented on asdf/commands/tests/test_extract.py:
```diff
@@ -1,11 +1,24 @@
+    # asdf.fits_embed is imported here
+    if "asdf.fits_embed" in sys.modules:
+        del sys.modules["asdf.fits_embed"]
+    # import of asdf.fits_embed is required here to allow for proper removal of
+    # the asdf.fits_embed module (see the above line of code) to trigger the
+    # deprecation warning on import for other tests
```
Given this comment, it seems like this module must be run first in order for the other tests to properly function? 

Tests should be independent of order of execution. If this needs to be done in this module then it must be done in every module expecting a similar warning so that they can be run independently.

braingram replied:
All of the fits_embed imports in the tests have a similar check to allow for independent order of execution.

WilliamJamieson replied:
I'm not taking about the check I'm talking about the `import asdf.fts_embed`.

braingram replied:
I changed the test to use `asdf.fits_embed` instead of `from asdf.fits_embed import AsdfInFits` and removed the comment (as the issues with removing the module caused by the `from` import are no longer relevant).

WilliamJamieson commented on asdf/commands/tests/test_remove_hdu.py:
```diff
@@ -1,10 +1,17 @@
-from asdf.fits_embed import AsdfInFits
+from asdf.exceptions import AsdfDeprecationWarning
+
+with pytest.warns(AsdfDeprecationWarning, match="AsdfInFits has been deprecated.*"):
+    if "asdf.fits_embed" in sys.modules:
+        del sys.modules["asdf.fits_embed"]
+    from asdf.fits_embed import AsdfInFits
```
Given all the discussion about `import asdf.fits_embed` shouldn't this be:
```suggestion
with pytest.warns(AsdfDeprecationWarning, match="AsdfInFits has been deprecated.*"):
    if "asdf.fits_embed" in sys.modules:
        del sys.modules["asdf.fits_embed"]
    import asdf.fitest_embed
    from asdf.fits_embed import AsdfInFits
```

braingram replied:
Similar to https://github.com/asdf-format/asdf/pull/1337#discussion_r1110074571 I changed the imports in this test to avoid the confusion.

WilliamJamieson commented:
> I am ok with this as a note and would prefer to have it reviewed.

I wanted to mark it as a draft to avoid an accidental merger until `sphinx_asdf` is released with a fix. I can still review things when it is a draft, it just won't be fully ready until the `sphinx_asdf` release.

braingram commented:
> I wanted to mark it as a draft to avoid an accidental merger until `sphinx_asdf` is released with a fix. I can still review things when it is a draft, it just won't be fully ready until the `sphinx_asdf` release.

@WilliamJamieson I believe I've addressed all of your comments except for this one. Am I correct in thinking that the inclusion of this warning (even as a note) means this PR has to wait for a release of `sphinx_asdf` before merging?

Would it be an acceptable alternative to include the deprecation description as a note and open an issue to convert it to a warning when `sphinx_asdf` is released?

WilliamJamieson commented:
> @WilliamJamieson I believe I've addressed all of your comments except for this one. Am I correct in thinking that the inclusion of this warning (even as a note) means this PR has to wait for a release of `sphinx_asdf` before merging?

Yes. Although, the only thing holding up a `sphinx_asdf` release are your reviews of

- asdf-format/sphinx-asdf#37
- asdf-format/sphinx-asdf#54

> 
> Would it be an acceptable alternative to include the deprecation description as a note and open an issue to convert it to a warning when `sphinx_asdf` is released?

This would be fine.

braingram commented:
> > @WilliamJamieson I believe I've addressed all of your comments except for this one. Am I correct in thinking that the inclusion of this warning (even as a note) means this PR has to wait for a release of `sphinx_asdf` before merging?
> 
> Yes. Although, the only thing holding up a `sphinx_asdf` release are your reviews of
> 
>     * [Move to pyproject.toml file sphinx-asdf#37](https://github.com/asdf-format/sphinx-asdf/pull/37)
> 
>     * [Move to using private block API sphinx-asdf#54](https://github.com/asdf-format/sphinx-asdf/pull/54)
> 

Those look unrelated to the removal of custom.css. Are you waiting for both of those before making a release? I doubt I will have a chance to review either before next week at the earliest.

> > Would it be an acceptable alternative to include the deprecation description as a note and open an issue to convert it to a warning when `sphinx_asdf` is released?
> 
> This would be fine.

Sounds great. I opened an issue to track changing the note back to a warning: https://github.com/asdf-format/asdf/issues/1386

WilliamJamieson commented:
@meekseeksdev backport 2.15.x

WilliamJamieson commented:
@meeseeksdev backport to 2.15.x
